### ISSUE TITLE
Allow changing of mysql credentials

### ISSUE THREAD

**Issue by bmschmidt** [senderle/bookworm-compose#10]
Mysql/mariadb SELECT queries are handled by granting select to 'bookworm@localhost'. But if these things are living on different docker services, is 'localhost' still going to work as part of the query?

Solution: probably will have make bookworm accept connections from any host, and trust the firewall.

**Comment by senderle** [comment]
Let me look at the way things are set up in the projects I've been running. I think the official mariadb image automatically expects external connections, so I have never actually had to think about this. And most setups I've seen just accept all connections (i.e. listen on 0.0.0.0), confident that those connections can only come from other services on the private network that `docker-compose` creates.

**Comment by senderle** [comment]
Following up on this briefly to note that if we only need one database and one user, this can be configured out of the box with the following environment variables (summarizing from https://hub.docker.com/_/mariadb):

MYSQL_ROOT_PASSWORD — we already know about
MYSQL_DATABASE — optional; the official image has a script that looks for this var and creates a database with this name at first startup
MYSQL_USER, MYSQL_PASSWORD — optional; if supplied, a user with the given name and password will be granted full access to the database created above

Thinking if we can cram the BookwormDB idiom into the above somehow, it would probably make some of these issues a lot simpler?

**Comment by bmschmidt** [comment]
The code is fairly insistent on having two users; an admin/root user to populate the databases, and then a read-only user deployed through the web host. There is no need for write access from the web server, so one of the major security walls has always been that the web service can never write anything.

It would probably be possible to hard-code into the image a command to create a select-only user.

It still matters--and I don't know--what port to tell the bookworm services to access it on. In general the code defaults to looking for the MySQL server on 127.0.0.1, which seems like it won't work here.

**Comment by bmschmidt** [comment]
OK, this is where I'm stuck now. Although I've put a mariadb client into the bookworm server, it's not clear to me yet just what credentials, hosts, and ports to use for remote login. In particular, I can't get into the mysql server explicitly invoking the password at '$MYSQL_ROOT_PASSWORD'
```
docker exec -it bookworm-compose_mariadb_1 /bin/bash
mysql #<- logs in as root
mysql -uroot -p
Enter password: <I enter the contents of echo $MYSQL_ROOT_PASSWORD>
ERROR 1045 (28000): Access denied for user 'root'@'localhost' (using password: YES)
```

**Comment by senderle** [comment]
Yeah that also looks like it should work. I will take a stab at troubleshooting this.

**Comment by senderle** [comment]
Ok, it's a persistence issue. Sorry, I forgot about this. Docker containers (noting the difference between _images_ and _containers_ here; if that's not clear let me know and I'll explain further, and OTOH sorry if this is all obvious already) are ephemeral by nature. They stick around sometimes but there are lots of events that can cause a container's state to be lost, which means you shouldn't store anything that needs to persist inside the container.

So that means the database files themselves get stored in a separate volume. But that volume persists sometimes too well! In this case it probably preserved an earlier password. The solution to your problem is to delete the volume and recreate it. First, tear down the stack, since otherwise docker will refuse to delete the in-use volume. Then remove the volume:

    docker-compose down
    docker volume rm mysql_data

Then recreate it

    docker volume create mysql_data

At this point in the process, I expected `docker-compose up` to work, but it still didn't, so there's one more problem.

It turned out that for some bizarre reason, the single quotes in `MYSQL_ROOT_PASSWORD='insecure_dev_root_password'` were being passed in as part of the password, which mariadb didn't like. I removed the single quotes and then everything worked. I don't 100% understand the quote conventions here; it's possible that using double quotes would have worked?? But anyway, in this case, it worked in this case without any quotes at all.

Sorry I keep forgetting about all these corner cases. Early stages of docker set-up are always kind of chaotic this way for me. But once they're over with, I almost never have to revisit them.

**Comment by bmschmidt** [comment]
OK, I've hacked this away in an insecure form. I couldn't see how MYSQL_ROOT_PASSWORD was working--it doesn't show up as ever having been defined in "SELECT * FROM mysql.user" inside the mariadb container--so I've just scripted `bin/run set_mysql_password insecure_dev_password` manually.

The next step would be to change that in a way keeps the values in `compose/bookworm/default_mysql_config/admin.cnf`--which is read only on the bookworm server--aligned with a changed password on the mariadb instance.
